File: CFSharp/CFSharp.F/MonadPolice.fs
module CFSharp.F.MonadPolice

open System.Linq
open FSharpx.Strings

type IEmailGateway =
    abstract Send : string -> string -> string -> string -> unit

type ImDave =
    abstract RecentRamblings : unit -> string seq // or Seq<string>
    
type MonadPolice(dave : ImDave, email : IEmailGateway) = 
    let _dave = dave
    let _email = email
    member this.Surveil() = 
        let overheard = _dave.RecentRamblings()
        let zealotTalk = overheard.Where(fun (x:string) -> x.Contains("monad"))
        for outburstOfZealotry in zealotTalk do
            _email.Send "xerxesb" "the monad police" "Lack of pragmatism detected" outburstOfZealotry

type MonadPolice_MoreFSharpish(dave : ImDave, email : IEmailGateway) = 
    let _dave = dave
    let _email = email
    let sendOutburst = _email.Send "xerxesb" "the monad police" "Lack of pragmatism detected"
    member this.Surveil() = 
        _dave.RecentRamblings()
            |> Seq.filter (contains "monad")
            |> Seq.iter sendOutburst

File: CFSharp/CFSharp.C/MonadPolice.cs
using System.Collections.Generic;
using System.Linq;

namespace CFSharp.C
{
    public interface IEmailGateway
    {
        void Send(string to, string from, string subject, string message);
    }

    public interface ImDave
    {
        IEnumerable<string> RecentRamblings();
    }

    public class MonadPolice
    {
        private readonly ImDave _dave;
        private readonly IEmailGateway _email;

        public MonadPolice(ImDave dave, IEmailGateway email)
        {
            _dave = dave;
            _email = email;
        }

        public void Surveil()
        {
            var overheard = _dave.RecentRamblings();

            var zealotTalk = overheard.Where(x => x.Contains("monad"));

            foreach (var outburstOfZealotry in zealotTalk)
            {
                _email.Send("xerxesb", "the monad police", "Lack of pragmatism detected", outburstOfZealotry);
            }
        }
    }
}
